Gameplay Stat Tracker V1

Sarge-117 opened the pull request:
This would introduce a feature to collect and display some stats about your playthrough of the game (whether vanilla or rando). The stats include showing chronological timestamps upon collecting major items and defeating bosses as well as keeping counters of various things, such as rupees collected/spent, number of rolls, sword swings, pots broken, bushes cut, chests opened, and a few others.

![coloured_timestamps](https://user-images.githubusercontent.com/108380086/202923765-2805b67f-33e3-4327-8c72-18eaa46794d5.PNG)

It also includes a count of enemies defeated which is broken down by individual enemy (this is the primary reason for the large number of files changed in this PR). Enemy counts are shown alphabetically in a collapsible tree, and only shows enemies that have been defeated at least once. Counts of ammo used and buttons pressed are similarly shown in collapsible trees.

Early footage showing timestamps:
https://cdn.discordapp.com/attachments/1019824570938699798/1039734068490805248/stat-tracking-test.mp4

Early footage showing enemies defeated/breakdown and ammo used:
https://cdn.discordapp.com/attachments/1034682552100855848/1041612176844136498/enemies_ammo.mp4

The stats are currently stored in `gSaveContext` and tied to the save file. For a "version 2" it would be nice to have an option to persist your stats without saving everything, but I feel that might be beyond me and this PR is already pretty large.

For future use, adding a new stat to be counted (or timestamp to be shown) is as simple as adding an entry to the corresponding enum in `gameplaystats.h` and then adding it to the lists in `gameplaystats.cpp`.

Shoutout to @RaelCappra for adding this to his experimental branch, which has allowed this to get some early testing.

Sarge-117 committed:
First test of gathering some gameplay stats

Sarge-117 committed:
Merge remote-tracking branch 'upstream/develop' into timer-stats-test

Sarge-117 committed:
Merge remote-tracking branch 'upstream/develop' into timer-stats-test

Sarge-117 committed:
timer changes and other stuff

Sarge-117 committed:
Merge remote-tracking branch 'upstream/develop' into timer-stats-test

Sarge-117 committed:
Move code to new files + rename

Sarge-117 committed:
Name change - gamePlayStats

Sarge-117 committed:
Finish rename, remove n64ddFlag checks

Sarge-117 committed:
Improve item get times

Sarge-117 committed:
Better time tracking, more stats,

Sarge-117 committed:
Put button under Enhancements

Sarge-117 committed:
Merge branch 'develop' into timer-stats-test

Sarge-117 committed:
Fix merge conflict

Sarge-117 committed:
Add pauseCount, fix bug with rando items

Sarge-117 committed:
Adjust inits/declarations

Sarge-117 committed:
step counter

Sarge-117 committed:
Name change: "itemGetTime" to "timestamp"

Sarge-117 committed:
Tidying + CI test

Sarge-117 committed:
Set up array for stat counts

Sarge-117 committed:
Macro

#define GAMEPLAYSTAT_TOTAL_TIME (gSaveContext.gameplayStats.playTimer / 2 + gSaveContext.gameplayStats.pauseTimer / 3)

Sarge-117 committed:
Add boss defeat timestamps

Sarge-117 committed:
Add sword swings, pots broken, bushes cut

Sarge-117 committed:
fix int type

Sarge-117 committed:
Merge branch 'develop' into timer-stats-test

Sarge-117 committed:
Merge remote-tracking branch 'upstream/develop' into timer-stats-test

Sarge-117 committed:
Add counts for enemies defeated

Broken down by enemy, with a total

Sarge-117 committed:
Add ammo used

Sarge-117 committed:
Hide breakdowns until count > 0

Sarge-117 committed:
Forgot Big Octo

Sarge-117 committed:
Count chests opened

Sarge-117 committed:
Merge branch 'develop' into timer-stats-test

Sarge-117 committed:
Update after LUS submodule

Sarge-117 committed:
Enemy count spacing

Sarge-117 committed:
Comments

Sarge-117 committed:
Merge branch 'develop' into timer-stats-test

Sarge-117 committed:
Count 3 mini Floormasters as 1 Floormaster

+ some cleanup

Sarge-117 committed:
Comments

Sarge-117 committed:
Colour coding for timestamps on quest items

i.e. medallions/stones/songs

Tulip-0333 reviewed (approved):
I love you (also for anyone seeing this I'm not underaged, please unban me from the Discord server.)

garrettjoecox commented:
Two initial thoughts, haven’t looked through the code much yet:
A) Why not put these into the SoHStats struct?
B) As it currently stands I’m not sure how I feel about the time stamp functionality being bundled with the rest of this. My thoughts on this aren’t fully developed yet, I’ll revisit later with some better feedback

Sarge-117 commented:
>A) Why not put these into the SoHStats struct?

I think I had this confused with the little stat window that shows the framerate + platform you're playing on. But good idea regardless, I'll make the switch.

Sarge-117 commented:
> A) Why not put these into the SoHStats struct?

I think I had this confused with the little stat window that shows the framerate + platform you're playing on. But good idea regardless, I'll make the switch.

Sarge-117 committed:
Move stat into the sohStats struct

+ rearrange the counts enum for easier addition of future counts

Sarge-117 committed:
Merge branch 'develop' into timer-stats-test

Archez commented:
My quick two cents: maybe the colors should be more UI friendly on a black screen, like use something that is easier on the eyes. The dark blue mostly is the one that stands out as hard to see for me.

Sarge-117 commented:
> My quick two cents: maybe the colors should be more UI friendly on a black screen, like use something that is easier on the eyes. The dark blue mostly is the one that stands out as hard to see for me.

I actually did brighten that one up somewhat after that screenshot was taken based on similar feedback from the experimental branch, so we should be good there!

Sarge-117 committed:
Some documentation + count button presses

Sarge-117 committed:
Stop counting button presses when Ganon defeated

Sarge-117 committed:
Couple bugfixes

Add count for Gerudo Thief, fix step counter counting in some situations where it shouldn't

Sarge-117 committed:
Fix comment

briaguya0 commented on soh/soh/SaveManager.cpp:
```diff
@@ -989,6 +998,14 @@ void SaveManager::LoadBaseVersion2() {
         SaveManager::Instance->LoadArray("dungeonKeys", ARRAY_COUNT(gSaveContext.sohStats.dungeonKeys), [](size_t i) {
             SaveManager::Instance->LoadData("", gSaveContext.sohStats.dungeonKeys[i]);
         });
+        SaveManager::Instance->LoadData("playTimer", gSaveContext.sohStats.playTimer);
+        SaveManager::Instance->LoadData("pauseTimer", gSaveContext.sohStats.pauseTimer);
+        SaveManager::Instance->LoadArray("timestamps", ARRAY_COUNT(gSaveContext.sohStats.timestamp), [](size_t i) {
+            SaveManager::Instance->LoadData("", gSaveContext.sohStats.timestamp[i]);
+        });
+        SaveManager::Instance->LoadArray("counts", ARRAY_COUNT(gSaveContext.sohStats.count), [](size_t i) {
+            SaveManager::Instance->LoadData("", gSaveContext.sohStats.count[i]);
+        });
```
will this cause any issues with loading old saves? i would assume it's fine considering these are new as opposed to modified, but it'd be nice to confirm

Sarge-117 replied:
I will try setting up some save files on 5.0.0 and see what happens loading them into this.

Sarge-117 replied:
Loading an older save doesn't seem to create any obvious problems! Obviously the timers and stats will be off for that particular save file but that's to be expected. @briaguya-ai

briaguya0 reviewed (approved):
this is great! after getting confirmation on that one question i left in a comment let's :shipit:
